lib/astc_encode.cpp

```diff
@@ -0,0 +1,925 @@
+                                                    KTX_FORMAT_VERSION_TWO);
+        levelImageSizeOut = ktxTexture_calcImageSize(ktxTexture(prototype), level,
+                                                     KTX_FORMAT_VERSION_TWO);
+        ktx_size_t offset = ktxTexture2_levelDataOffset(This, level);
+
+        for (uint32_t image = 0; image < levelImages; image++) {
+            if (params->verbose)
+                std::cout << "ASTC compressor: compressing image " <<
+                             (This->numLevels - level - 1) * levelImages + image + 1
+                             << " of " << This->numLevels * levelImages
+                             << std::endl;
+
+            astcenc_image *input_image = nullptr;
+            if (num_components == 1)
+                input_image = unorm8x1ArrayToImage(This->pData + offset,
+                                                   width, height);
+            else if (num_components == 2)
+                input_image = unorm8x2ArrayToImage(This->pData + offset,
+                                                   width, height);
+            else if (num_components == 3)
+                input_image = unorm8x3ArrayToImage(This->pData + offset,
+                                                   width, height);
+            else // assume (num_components == 4)
+                input_image = unorm8x4ArrayToImage(This->pData + offset,
+                                                   width, height);
+
+            assert(input_image);
+
+            CompressionWorkload work;
+            work.context = astc_context;
+            work.image = input_image;
+            work.swizzle = swizzle;
+            work.data_out = buffer_out;
+            work.data_len = levelImageSizeOut;
+            work.error = ASTCENC_SUCCESS;
+
+            launchThreads(threadCount, compressionWorkloadRunner, &work);
+
+            if (work.error != ASTCENC_SUCCESS) {
+                std::cout << "ASTC compressor failed\n" <<
+                             astcenc_get_error_string(work.error) << std::endl;
+
+                imageFree(input_image);
+
+                astcenc_context_free(astc_context);
+                return KTX_INVALID_OPERATION;
+            }
+
+            buffer_out += levelImageSizeOut;
+
+            // Reset ASTC context for next image
+            astcenc_compress_reset(astc_context);
+            offset += levelImageSizeIn;
+        }
+    }
+
+    // We are done with astcencoder
+    astcenc_context_free(astc_context);
+
+    // reset pointer to new BDB
+    BDB = prototype->pDfd+1;
+
+    // ASTC-related checks
+    ktx_uint32_t model = KHR_DFDVAL(BDB, MODEL);
+    ktx_uint32_t primaries = KHR_DFDVAL(BDB, PRIMARIES);
+
+    assert(model == KHR_DF_MODEL_ASTC && "Invalid dfd generated for ASTC image\n");
+    if (transfer == KHR_DF_TRANSFER_SRGB) {
+        assert(primaries == KHR_DF_PRIMARIES_SRGB && "Not a valid sRGB image\n");
+    }
+
```

Comment:
This is going to cause compile warnings and possibly errors in the Release config as there will be no asserts. I suggest

```c++
assert(KHR_DFDVAL(prototype->pDfd+1, MODEL) == KHR_DF_MODEL_ASTC
       && "Invalid dfd generated for ASTC image\n");
assert((transfer == KHR_DF_TRANSFER_SRGB 
       ? KHR_DFDVAL(prototype->pDfd+1, PRIMARIES) == KHR_DF_PRIMARIES_SRGB
       : true) && "Not a valid sRGB image\n");
```

Reply:
These are fixed in later commits :). I think since we have lots of comments and quite a few commits now. This might be very confusing.

Reply:
I have "changes from all commits" selected and I'm still seeing the lines as I saw them when I made the comment.

Reply:
Apologies my bad. I made the change last nite once I saw the warnings from CI. But haven't commited the code yet.

Reply:
Fixed in https://github.com/KhronosGroup/KTX-Software/pull/433/commits/18af4983456764eb784babe2d11a125f5d2b7f5c